In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

from sklearn.model_selection import train_test_split
from sklearn.metrics.pairwise import cosine_similarity
import re


In [3]:
books = pd.read_csv('datasets/Books.csv')
users = pd.read_csv("datasets/Users.csv")
rating = pd.read_csv("datasets/Ratings.csv")

/tmp/ipykernel_12880/4095179287.py:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv('datasets/Books.csv')


In [4]:
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [5]:
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [6]:
rating.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [7]:
print("Books:",books.shape,"Users:",users.shape,"Ratings:",rating.shape)

Books: (271360, 8) Users: (278858, 3) Ratings: (1149780, 3)


In [8]:
books.isna().sum()

ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [9]:
users.isna().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

In [10]:
rating.isna().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

From Books:
Few missing values, they can be dropped(for object or string values) or can be imputed with mode(for numerical values).

From Users:
A lot of age is missing and age is not important for this project so we will be dropping age column.

From Ratings:
No missing values at all.

In [11]:
users.dropna(axis=1, inplace=True)
print(users.shape)
users.head()

(278858, 2)


User-ID                            Location
0        1                  nyc, new york, usa
1        2           stockton, california, usa
2        3     moscow, yukon territory, russia
3        4           porto, v.n.gaia, portugal
4        5  farnborough, hants, united kingdom

In [12]:
books.duplicated().sum()

0

In [13]:
rating.duplicated().sum()

0

In [14]:
users.duplicated().sum()

0

## EDA

## Popularity Based Filtering

In [15]:
book_ratings = rating.merge(books,on="ISBN")
book_ratings.head()

User-ID        ISBN  Book-Rating            Book-Title Book-Author  \
0   276725  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
1     2313  034545104X            5  Flesh Tones: A Novel  M. J. Rose   
2     6543  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
3     8680  034545104X            5  Flesh Tones: A Novel  M. J. Rose   
4    10314  034545104X            9  Flesh Tones: A Novel  M. J. Rose   

  Year-Of-Publication         Publisher  \
0                2002  Ballantine Books   
1                2002  Ballantine Books   
2                2002  Ballantine Books   
3                2002  Ballantine Books   
4                2002  Ballantine Books   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/034545104X.0...  
1  http://images.amazon.com/images/P/034545104X.0...  
2  http://images.amazon.com/images/P/034545104X.0...  
3  http://images.amazon.com/images/P/034545104X.0...  
4  http://images.amazon.com/images/P/034545104X.0...

In [16]:
rating_counts = book_ratings.groupby("Book-Title").count()['Book-Rating'].reset_index()
rating_counts.head(3)

Book-Title  Book-Rating
0   A Light in the Storm: The Civil War Diary of ...            4
1                              Always Have Popsicles            1
2               Apple Magic (The Collector's series)            1

In [17]:
rating_avg = book_ratings.groupby("Book-Title").mean()['Book-Rating'].reset_index()
rating_avg.head(3)

/tmp/ipykernel_12880/2074531782.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  rating_avg = book_ratings.groupby("Book-Title").mean()['Book-Rating'].reset_index()


Book-Title  Book-Rating
0   A Light in the Storm: The Civil War Diary of ...         2.25
1                              Always Have Popsicles         0.00
2               Apple Magic (The Collector's series)         0.00

In [18]:
rated_books = rating_counts.merge(rating_avg,on="Book-Title")
rated_books.head(5)

Book-Title  Book-Rating_x  \
0   A Light in the Storm: The Civil War Diary of ...              4   
1                              Always Have Popsicles              1   
2               Apple Magic (The Collector's series)              1   
3   Ask Lily (Young Women of Faith: Lily Series, ...              1   
4   Beyond IBM: Leadership Marketing and Finance ...              1   

   Book-Rating_y  
0           2.25  
1           0.00  
2           0.00  
3           8.00  
4           0.00

In [19]:
rated_books.rename(columns={'Book-Rating_x':"Rating_Count", 'Book-Rating_y':"Avg_Ratings"},inplace=True)
rated_books.head()

Book-Title  Rating_Count  \
0   A Light in the Storm: The Civil War Diary of ...             4   
1                              Always Have Popsicles             1   
2               Apple Magic (The Collector's series)             1   
3   Ask Lily (Young Women of Faith: Lily Series, ...             1   
4   Beyond IBM: Leadership Marketing and Finance ...             1   

   Avg_Ratings  
0         2.25  
1         0.00  
2         0.00  
3         8.00  
4         0.00

In [20]:
# Storing 100 top rated books in rated_Books
rated_Books = rated_books[rated_books['Rating_Count']>=250].sort_values('Avg_Ratings',ascending=False).head(100)

In [48]:
popular_books = rated_Books.merge(books,on="Book-Title").drop_duplicates('Book-Title')[['Book-Title','Book-Author','Year-Of-Publication','Image-URL-M','Image-URL-L','Rating_Count','Avg_Ratings']]

In [52]:
popular_books.isna().sum()

Book-Title             0
Book-Author            0
Year-Of-Publication    0
Image-URL-M            0
Image-URL-L            0
Rating_Count           0
Avg_Ratings            0
dtype: int64

In [49]:
popular_books.head()


Book-Title    Book-Author  \
0   Harry Potter and the Prisoner of Azkaban (Book 3)  J. K. Rowling   
3        Harry Potter and the Goblet of Fire (Book 4)  J. K. Rowling   
5      Harry Potter and the Sorcerer's Stone (Book 1)  J. K. Rowling   
9   Harry Potter and the Order of the Phoenix (Boo...  J. K. Rowling   
13   Harry Potter and the Chamber of Secrets (Book 2)  J. K. Rowling   

   Year-Of-Publication                                        Image-URL-M  \
0                 1999  http://images.amazon.com/images/P/0439136350.0...   
3                 2000  http://images.amazon.com/images/P/0439139597.0...   
5                 1998  http://images.amazon.com/images/P/0590353403.0...   
9                 2003  http://images.amazon.com/images/P/043935806X.0...   
13                2000  http://images.amazon.com/images/P/0439064872.0...   

                                          Image-URL-L  Rating_Count  \
0   http://images.amazon.com/images/P/0439136350.0...           428   
3   http://images.amazon.com/images/P/0439139597.0...           387   
5   http://images.amazon.com/images/P/0590353403.0...           278   
9   http://images.amazon.com/images/P/043935806X.0...           347   
13  http://images.amazon.com/images/P/0439064872.0...           556   

    Avg_Ratings  
0      5.852804  
3      5.824289  
5      5.737410  
9      5.501441  
13     5.183453

## Collaborative Filtering

(Users on columns and books on index and the data tells how much that user rated that book)

1. Users = users.rating > 200
2. books = books.times_rated > 50

In [23]:
users_rate_200 = book_ratings.groupby('User-ID').count()['Book-Rating'] > 200
users_rate_idx = users_rate_200[users_rate_200].index

In [24]:
filtered_users = book_ratings[book_ratings['User-ID'].isin(users_rate_idx)]
filtered_users

User-ID        ISBN  Book-Rating  \
2           6543  034545104X            0   
5          23768  034545104X            0   
7          28523  034545104X            0   
15         77940  034545104X            0   
16         81977  034545104X            0   
...          ...         ...          ...   
1030883   275970  1880837927            0   
1030884   275970  188717897X            0   
1030885   275970  1888889047            0   
1030886   275970  1931868123            0   
1030887   275970  3411086211           10   

                                                Book-Title  \
2                                     Flesh Tones: A Novel   
5                                     Flesh Tones: A Novel   
7                                     Flesh Tones: A Novel   
15                                    Flesh Tones: A Novel   
16                                    Flesh Tones: A Novel   
...                                                    ...   
1030883                         The Theology of the Hammer   
1030884  The Ordeal of Integration: Progress and Resent...   
1030885  Pushcart's Complete Rotten Reviews &amp; Rejec...   
1030886  There's a Porcupine in My Outhouse: Misadventu...   
1030887                                         Die Biene.   

                           Book-Author Year-Of-Publication  \
2                           M. J. Rose                2002   
5                           M. J. Rose                2002   
7                           M. J. Rose                2002   
15                          M. J. Rose                2002   
16                          M. J. Rose                2002   
...                                ...                 ...   
1030883                 Millard Fuller                1994   
1030884              Orlando Patterson                1998   
1030885                 Bill Henderson                1998   
1030886                   Mike Tougias                2002   
1030887  Sybil Gr�¤fin Sch�¶nfeldt                1993   

                                    Publisher  \
2                            Ballantine Books   
5                            Ballantine Books   
7                            Ballantine Books   
15                           Ballantine Books   
16                           Ballantine Books   
...                                       ...   
1030883         Smyth &amp; Helwys Publishing   
1030884                Civitas Book Publisher   
1030885                        Pushcart Press   
1030886                    Capital Books (VA)   
1030887  Bibliographisches Institut, Mannheim   

                                               Image-URL-S  \
2        http://images.amazon.com/images/P/034545104X.0...   
5        http://images.amazon.com/images/P/034545104X.0...   
7        http://images.amazon.com/images/P/034545104X.0...   
15       http://images.amazon.com/images/P/034545104X.0...   
16       http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1030883  http://images.amazon.com/images/P/1880837927.0...   
1030884  http://images.amazon.com/images/P/188717897X.0...   
1030885  http://images.amazon.com/images/P/1888889047.0...   
1030886  http://images.amazon.com/images/P/1931868123.0...   
1030887  http://images.amazon.com/images/P/3411086211.0...   

                                               Image-URL-M  \
2        http://images.amazon.com/images/P/034545104X.0...   
5        http://images.amazon.com/images/P/034545104X.0...   
7        http://images.amazon.com/images/P/034545104X.0...   
15       http://images.amazon.com/images/P/034545104X.0...   
16       http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1030883  http://images.amazon.com/images/P/1880837927.0...   
1030884  http://images.amazon.com/images/P/188717897X.0...   
1030885  http://images.amazon.com/images/P/1888889047.0...   
1030886  http://images.a

In [25]:
y = filtered_users.groupby('Book-Title').count()['Book-Rating'] > 50
books_idx = y[y].index

In [26]:
final_filtered_df = filtered_users[filtered_users['Book-Title'].isin(books_idx)]
print(final_filtered_df.shape)
final_filtered_df.head()


(57236, 10)


User-ID        ISBN  Book-Rating    Book-Title      Book-Author  \
63   278418  0446520802            0  The Notebook  Nicholas Sparks   
65     3363  0446520802            0  The Notebook  Nicholas Sparks   
66     7158  0446520802           10  The Notebook  Nicholas Sparks   
69    11676  0446520802           10  The Notebook  Nicholas Sparks   
74    23768  0446520802            6  The Notebook  Nicholas Sparks   

   Year-Of-Publication     Publisher  \
63                1996  Warner Books   
65                1996  Warner Books   
66                1996  Warner Books   
69                1996  Warner Books   
74                1996  Warner Books   

                                          Image-URL-S  \
63  http://images.amazon.com/images/P/0446520802.0...   
65  http://images.amazon.com/images/P/0446520802.0...   
66  http://images.amazon.com/images/P/0446520802.0...   
69  http://images.amazon.com/images/P/0446520802.0...   
74  http://images.amazon.com/images/P/0446520802.0...   

                                          Image-URL-M  \
63  http://images.amazon.com/images/P/0446520802.0...   
65  http://images.amazon.com/images/P/0446520802.0...   
66  http://images.amazon.com/images/P/0446520802.0...   
69  http://images.amazon.com/images/P/0446520802.0...   
74  http://images.amazon.com/images/P/0446520802.0...   

                                          Image-URL-L  
63  http://images.amazon.com/images/P/0446520802.0...  
65  http://images.amazon.com/images/P/0446520802.0...  
66  http://images.amazon.com/images/P/0446520802.0...  
69  http://images.amazon.com/images/P/0446520802.0...  
74  http://images.amazon.com/images/P/0446520802.0...

In [27]:
# Creating desired df of users and books using pivot table
pt = final_filtered_df.pivot_table(index='Book-Title', columns='User-ID',values='Book-Rating')
print(pt.shape)
pt.head()

(679, 810)


User-ID              254     2276    2766    2977    3363    4017    4385    \
Book-Title                                                                    
1984                    9.0     NaN     NaN     NaN     NaN     NaN     NaN   
1st to Die: A Novel     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2nd Chance              NaN    10.0     NaN     NaN     NaN     NaN     NaN   
4 Blondes               NaN     NaN     NaN     NaN     NaN     NaN     NaN   
A Bend in the Road      0.0     NaN     7.0     NaN     NaN     NaN     NaN   

User-ID              6251    6323    6543    ...  271705  273979  274004  \
Book-Title                                   ...                           
1984                    NaN     NaN     NaN  ...    10.0     NaN     NaN   
1st to Die: A Novel     NaN     NaN     9.0  ...     NaN     NaN     NaN   
2nd Chance              NaN     NaN     0.0  ...     NaN     NaN     NaN   
4 Blondes               0.0     NaN     NaN  ...     NaN     NaN     NaN   
A Bend in the Road      NaN     NaN     NaN  ...     NaN     0.0     NaN   

User-ID              274061  274301  274308  275970  277427  277639  278418  
Book-Title                                                                   
1984                    NaN     NaN     NaN     0.0     NaN     NaN     NaN  
1st to Die: A Novel     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
2nd Chance              NaN     NaN     0.0     NaN     NaN     0.0     NaN  
4 Blondes               NaN     NaN     NaN     NaN     NaN     NaN     NaN  
A Bend in the Road      NaN     NaN     NaN     NaN     NaN     NaN     NaN  

[5 rows x 810 columns]

In [28]:
pt.fillna(0,inplace=True)
pt.head()

User-ID              254     2276    2766    2977    3363    4017    4385    \
Book-Title                                                                    
1984                    9.0     0.0     0.0     0.0     0.0     0.0     0.0   
1st to Die: A Novel     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
2nd Chance              0.0    10.0     0.0     0.0     0.0     0.0     0.0   
4 Blondes               0.0     0.0     0.0     0.0     0.0     0.0     0.0   
A Bend in the Road      0.0     0.0     7.0     0.0     0.0     0.0     0.0   

User-ID              6251    6323    6543    ...  271705  273979  274004  \
Book-Title                                   ...                           
1984                    0.0     0.0     0.0  ...    10.0     0.0     0.0   
1st to Die: A Novel     0.0     0.0     9.0  ...     0.0     0.0     0.0   
2nd Chance              0.0     0.0     0.0  ...     0.0     0.0     0.0   
4 Blondes               0.0     0.0     0.0  ...     0.0     0.0     0.0   
A Bend in the Road      0.0     0.0     0.0  ...     0.0     0.0     0.0   

User-ID              274061  274301  274308  275970  277427  277639  278418  
Book-Title                                                                   
1984                    0.0     0.0     0.0     0.0     0.0     0.0     0.0  
1st to Die: A Novel     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
2nd Chance              0.0     0.0     0.0     0.0     0.0     0.0     0.0  
4 Blondes               0.0     0.0     0.0     0.0     0.0     0.0     0.0  
A Bend in the Road      0.0     0.0     0.0     0.0     0.0     0.0     0.0  

[5 rows x 810 columns]

**Each book can be seen as a vector. A vector of 810 numbers. 
 Using cosine similarity to check the similarity between these all vector representations of books.**


In [29]:
similarity_score = cosine_similarity(pt)
similarity_score.shape


(679, 679)

**This gives a similarity score between each book to other book. 
There are 679 books and each array (say similarity_score[0]) gives its similarity with all 679 books**

### Recommendation

Let's build a function that recommends us 5 best matching suggestion for book.

In [30]:
# index = np.where(pt.index=='You Belong To Me')[0][0]
# index

In [31]:
# similarity_score[index]

In [32]:
# list(enumerate(similarity_score[index]))

In [33]:
# sorted(list(enumerate(similarity_score[index])),key=lambda x: x[1],reverse=True)[1:6]

In [34]:
def find_best_match(input_book_name):
    best_match = None
    best_match_score = 0

    input_words = set(re.findall(r'\b\w+\b', input_book_name.lower()))

    for book_name in pt.index:
        book_words = set(re.findall(r'\b\w+\b', book_name.lower()))

        # Calculate the similarity score by finding the intersection of words
        match_score = len(input_words.intersection(book_words))

        if match_score > best_match_score:
            best_match = book_name
            best_match_score = match_score

    return best_match

In [35]:
def recommend(book_name):
    assert type(book_name)==str
    if book_name in pt.index:
        index = np.where(pt.index==book_name)[0][0]
        similar_books = sorted(list(enumerate(similarity_score[index])),key=lambda x: x[1],reverse=True)[1:6]
        
        for i in similar_books:
            print(pt.index[i[0]])
    else:
        print("Your provided Book Name didn't match. Searching for best match:")
        book = find_best_match(book_name)
        print("Searching for\t:", book)
        print("")
        if book != None:
            recommend(book)
        else:
            print("Try for another book")

            

In [36]:
recommend("Da vinci")

Your provided Book Name didn't match. Searching for best match:
Searching for	: The Da Vinci Code

Angels &amp; Demons
Touching Evil
Saving Faith
The Sweet Potato Queens' Book of Love
Middlesex: A Novel


In [37]:
find_best_match("Da Vinci")

'The Da Vinci Code'

In [38]:
recommend('You Belong To Me')


Loves Music, Loves to Dance
I'll Be Seeing You
Before I Say Good-Bye
Daddy's Little Girl
All Around the Town


In [39]:
recommend('Zoya')

Fine Things
Secrets
Kaleidoscope
Exclusive
Five Days in Paris


In [40]:
recommend('Message in a Bottle')

Nights in Rodanthe
The Mulberry Tree
A Walk to Remember
River's End
Nightmares &amp; Dreamscapes


In [41]:
recommend("The Da Vinci Code")

Angels &amp; Demons
Touching Evil
Saving Faith
The Sweet Potato Queens' Book of Love
Middlesex: A Novel


In [42]:
recommend("Notebook")

Your provided Book Name didn't match. Searching for best match:
Searching for	: The Notebook

A Walk to Remember
The Rescue
One Door Away from Heaven
Toxin
The Five People You Meet in Heaven


In [43]:
recommend("Remember")

Your provided Book Name didn't match. Searching for best match:
Searching for	: A Walk to Remember

The Rescue
Nights in Rodanthe
The Notebook
Granny Dan
A Bend in the Road


In [44]:
recommend("Harry Potter")

Your provided Book Name didn't match. Searching for best match:
Searching for	: Harry Potter and the Chamber of Secrets (Book 2)

Harry Potter and the Prisoner of Azkaban (Book 3)
Harry Potter and the Goblet of Fire (Book 4)
Harry Potter and the Sorcerer's Stone (Harry Potter (Paperback))
Harry Potter and the Sorcerer's Stone (Book 1)
Harry Potter and the Order of the Phoenix (Book 5)


In [45]:
recommend("Hobbit")

Your provided Book Name didn't match. Searching for best match:
Searching for	: The Hobbit : The Enchanting Prelude to The Lord of the Rings

The Two Towers (The Lord of the Rings, Part 2)
The Fellowship of the Ring (The Lord of the Rings, Part 1)
Where the Red Fern Grows
One for the Money (A Stephanie Plum Novel)
The Return of the King (The Lord of the Rings, Part 3)


In [51]:
import pickle
pickle.dump(popular_books,open('popular_books.pkl','wb'))